In [23]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [24]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df_lending = pd.read_csv("lending_data.csv")

# Review the DataFrame
df_lending.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [25]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending["loan_status"].values.reshape(-1, 1)

# Separate the X variable, the features
X = df_lending.copy()
X.drop("loan_status", axis=1, inplace=True)



In [26]:
# Review the y variable Series
y[:5]

array([[0],
       [0],
       [0],
       [0],
       [0]])

In [27]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [28]:
# Check the balance of our target values
unique, counts = np.unique(y, return_counts=True)
print(np.asarray((unique, counts)).T)


[[    0 75036]
 [    1  2500]]


### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [29]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [30]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
lr_train_model = LogisticRegression()

# Fit the model using training data
lr_train_model.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression()

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [31]:
# Make a prediction using the testing data
y_pred_test = lr_train_model.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [32]:
# Print the balanced_accuracy score of the model
print(balanced_accuracy_score(y_test, y_pred_test))

0.9481269535981227


In [33]:
# Generate a confusion matrix for the model
cm = confusion_matrix(y_test, y_pred_test)
cm_df = pd.DataFrame(
    cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)
display(cm_df)


Predicted 0  Predicted 1
Actual 0        22347          121
Actual 1           78          715

In [ ]:
on size,

In [34]:
# Print the classification report for the model
target_names=['Healthy loan', 'High-risk loan']
print(classification_report(y_test, y_pred_test, target_names=target_names))


                precision    recall  f1-score   support

  Healthy loan       1.00      0.99      1.00     22468
High-risk loan       0.86      0.90      0.88       793

      accuracy                           0.99     23261
     macro avg       0.93      0.95      0.94     23261
  weighted avg       0.99      0.99      0.99     23261



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Accuracy of the model was at 94.8%. However, prediction for helthy loans was 100%, but high-risk was at 86%

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points.

In [35]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
oversample = RandomOverSampler(random_state=1)


# Fit the original training data to the random_oversampler model
X_resample, y_resample = oversample.fit_resample(X_train, y_train)

In [36]:
X_resample

loan_size  interest_rate  borrower_income  debt_to_income  \
0          9900.0          7.331            49600        0.395161   
1          9300.0          7.093            47300        0.365751   
2          8300.0          6.658            43300        0.307159   
3          8600.0          6.761            44200        0.321267   
4          9100.0          7.001            46500        0.354839   
...           ...            ...              ...             ...   
105131    20000.0         11.629            90000        0.666667   
105132    20100.0         11.647            90200        0.667406   
105133    20800.0         11.948            93000        0.677419   
105134    20200.0         11.716            90900        0.669967   
105135    20200.0         11.692            90600        0.668874   

        num_of_accounts  derogatory_marks  total_debt  
0                     4                 0       19600  
1                     3                 0       17300  
2                     2                 0       13300  
3                     3                 0       14200  
4                     3                 0       16500  
...                 ...               ...         ...  
105131               13                 3       60000  
105132               13                 3       60200  
105133               14                 3       63000  
105134               13                 3       60900  
105135               13                 3       60600  

[105136 rows x 7 columns]

In [37]:
# Count the distinct values of the resampled labels data
unique, counts = np.unique(y_resample, return_counts=True)
print(np.asarray((unique, counts)).T)

[[    0 52568]
 [    1 52568]]


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [38]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
resample_model = LogisticRegression()


# Fit the model using the resampled training data
resample_model.fit(X_resample, y_resample)

# Make a prediction using the testing data
y_pred_sample = resample_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [39]:
# Print the balanced_accuracy score of the model
print(balanced_accuracy_score(y_test, y_pred_sample))

0.9931903712406109


In [18]:
# Generate a confusion matrix for the model
cm_sample = confusion_matrix(y_test, y_pred_sample)
cm_sample_df = pd.DataFrame(
    cm_sample, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)
display(cm_sample_df)


Predicted 0  Predicted 1
Actual 0        22389          131
Actual 1            2          739

In [40]:
# Print the classification report for the model
print(classification_report(y_test, y_pred_sample,target_names=target_names))

                precision    recall  f1-score   support

  Healthy loan       1.00      0.99      1.00     22468
High-risk loan       0.85      0.99      0.92       793

      accuracy                           0.99     23261
     macro avg       0.93      0.99      0.96     23261
  weighted avg       0.99      0.99      0.99     23261



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** This model was able to predict healthy loans at 100%, but high-risk loans at 85%.